lib/procore/http_client/oauth.ex

```diff
@@ -1,62 +1,59 @@
 defmodule Procore.HttpClient.OAuth do
-  @doc """
+  @moduledoc """
   Hits OAuth endpoint and returns a token.
 
   Currently, options such as url, id, secret are set
   using env variables in config.exs
-
-  ## Examples
-  iex> get_oauth_token()
-  "asldkfjdasofiujao324qf832jsakasdlfkjafldsalf"
   """
-  @spec get_oauth_token() :: binary()
-  def get_oauth_token() do
-    url = Application.get_env(:procore, :oauth_url, "https://api.procore.com/oauth/token")
-    id = Application.get_env(:procore, :client_id)
-    secret = Application.get_env(:procore, :client_secret)
 
-    get_oauth_token(url, id, secret)
-  end
+  use Tesla
+  plug(Tesla.Middleware.EncodeJson, engine: Poison)
+  plug(Tesla.Middleware.Logger)
 
-  defp get_oauth_token(url, id, secret) do
-    cache_key = "OAUTH:#{url}:#{id}:#{secret}"
+  @url Application.get_env(:procore, :oauth_url, "https://api.procore.com/oauth/token")
+  @id Application.get_env(:procore, :client_id)
+  @secret Application.get_env(:procore, :client_secret)
+
+  @spec get_oauth_token() :: binary()
+  def get_oauth_token() do
+    cache_key = "OAUTH:#{@url}:#{@id}:#{@secret}"
 
     case Cachex.get(:a, cache_key) do
-      {:ok, oauth_oken} ->
-        oauth_oken
+      {:ok, oauth_token} ->
+        oauth_token
 
       {:missing, _} ->
-        {body, headers} = oauth_body_headers(id, secret)
-
-        case HTTPoison.post(url, body, headers, []) do
-          {:ok, %HTTPoison.Response{status_code: 200, body: body}} ->
+        case post(tesla_client(), @url, auth_body()) do
+          {:ok, %Tesla.Env{status: 200, body: body}} ->
             cache_and_return_token(body, cache_key)
 
-          {:ok, %HTTPoison.Response{status_code: 404}} ->
+          {:ok, %Tesla.Env{status: 404}} ->
             IO.puts("Not found :(")
 
-          {:error, %HTTPoison.Error{reason: reason}} ->
+          {:ok, %Tesla.Env{status: 401}} ->
+            IO.puts("Bad Credentials")
+
+          {:error, reason} ->
             IO.inspect(reason)
         end
     end
   end
 
-  defp oauth_body_headers(id, secret) do
-    body =
-      Poison.encode!(%{
-        client_id: id,
-        client_secret: secret,
-        grant_type: "client_credentials"
-      })
-
-    headers = [{"Content-type", "application/json"}]
-
-    {body, headers}
+  defp auth_body do
+    %{client_id: @id, client_secret: @secret, grant_type: "client_credentials"}
   end
 
   defp cache_and_return_token(body, cache_key) do
     %{"access_token" => token, "expires_in" => expires_in} = Poison.decode!(body)
     Cachex.set(:a, cache_key, token, ttl: :timer.seconds(expires_in - 1000))
     token
   end
+
+  defp tesla_client do
+    Tesla.build_client([{Tesla.Middleware.Headers, headers()}])
+  end
+
+  defp headers do
+    [{"Content-Type", "application/json"}]
+  end
 end
```

Comment:
I wrote a plug to do oauth that might be more elegant than this, assuming it actually fits into the existing architecture. If you are interested, you can find it here: https://github.com/procore/project-mirror/blob/master/lib/project_mirror/procore/oauth_middleware.ex

Reply:
Will definitely take a look at this and fit it into the sdk. We haven't focused much on auth, so I'm sure the middleware will be an improvement.